src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Windows.cs

```diff
@@ -56,8 +56,417 @@ private static unsafe void ThrowExceptionEncryptDecryptFail(string fullPath)
+                        }
+                        if (destinationHandle.IsInvalid)
+                        {
+                            // Try opening as a new file.
+                            destinationHandle = null;
+                        }
+                    }
+
+                    // If we still haven't opened the file.
+                    if (destinationHandle == null)
+                    {
+                        // Try to create the destination file.
+                        destinationHandle = TryOpenTargetNoAlternativeDataStream(
+                            destFullPath,
+                            Interop.Kernel32.GenericOperations.GENERIC_READ | Interop.Kernel32.GenericOperations.GENERIC_WRITE,
+                            FileShare.None,
+                            FileMode.CreateNew,
+                            0,
+                            ref destFileInformation,
+                            out _);
+                        if (destinationHandle == null)
+                        {
+                            // Alternative data stream.
+                            return false;
+                        }
+                        if (destinationHandle.IsInvalid)
+                        {
+                            // Failure to open.
+                            return false;
+                        }
+                        madeNew = true;
+                    }
+
+                    // Read the destination file's volume's serial number.
+                    int destSerialNumber;
+                    if (Interop.Kernel32.GetVolumeInformationByHandle(sourceHandle, out _, false, &destSerialNumber, null, out int destVolumeFlags, null, 0) != 0)
+                    {
+                        throw new Exception("G");
+                        //return false;
+                    }
+
+                    // Check the destination file is on the same volume.
+                    // Note: this doesn't guarantee they're on the same volume for sure, so we have to take that into consideration later: https://devblogs.microsoft.com/oldnewthing/20170707-00/?p=96555.
+                    if (sourceSerialNumber != destSerialNumber)
+                    {
+                        throw new Exception("H");
+                        //return false;
+                    }
+
+                    // Quick sanity check to see if the destination supports the block copy operation, since it's basically free to check anyway.
+                    if ((destVolumeFlags & Interop.Kernel32.FILE_SUPPORTS_BLOCK_REFCOUNTING) == 0)
+                    {
+                        throw new Exception("I");
+                        //return false;
+                    }
+
+                    // Get the source volume path. Note: we need the real path here for symlinks also, hence openedSourceFullPath.
+                    if (Interop.Kernel32.GetVolumePathName(openedSourceFullPath, out var volumePath) != 0)
```

Comment:
If we could avoid needing `openedSourceFullPath` here that'd be great. This would also be an issue if we used `FindFirstStreamW`. If we could avoid needing the actual source path (`openedSourceFullPath`) (ie. the actual path we have opened as a handle, with symlinks resolved etc.) we could easily delay 1 syscall in the golden path to failure, by being able to avoid checking attributes to see if it's a symlink.